src/dodal/devices/zocalo/zocalo_results.py

```diff
@@ -0,0 +1,224 @@
+        prefix: str = "",
+    ) -> None:
+        self.zocalo_environment = zocalo_environment
+        self.sort_key = SortKeys[sort_key]
+        self.channel = channel
+        self.timeout_s = timeout_s
+        self._prefix = prefix
+
+        self._subscription_run: bool = False
+        self._raw_results_received: asyncio.Queue = asyncio.Queue()
+
+        self.results = create_soft_signal_r(list[XrcResult], "results", self.name)
+        self.centres_of_mass = create_soft_signal_r(
+            NDArray[np.uint64], "centres_of_mass", self.name
+        )
+        self.bbox_sizes = create_soft_signal_r(
+            NDArray[np.uint64], "bbox_sizes", self.name
+        )
+        self.set_readable_signals(
+            read=[
+                self.results,
+                self.centres_of_mass,
+                self.bbox_sizes,
+            ]
+        )
+        super().__init__(name)
+
+    async def _put_results(self, results: Sequence[XrcResult]):
+        await self.results._backend.put(list(results))
+        centres_of_mass = np.array([r["centre_of_mass"] for r in results])
+        bbox_sizes = np.array([bbox_size(r) for r in results])
+        await self.centres_of_mass._backend.put(centres_of_mass)
+        await self.bbox_sizes._backend.put(bbox_sizes)
+
+    @AsyncStatus.wrap
+    async def trigger(self):
+        """Returns an AsyncStatus waiting for results to be received from Zocalo."""
+        LOGGER.info("Zocalo trigger called")
+        if not self._subscription_run:
+            LOGGER.info("subscription not initialised, subscribing to queue")
+            self._subscribe_to_results()
+            self._subscription_run = True
+
+        try:
+            LOGGER.info("waiting for results in queue")
+
+            async def _get_results():
+                raw_results = await self._raw_results_received.get()
+                LOGGER.info(f"Zocalo: found {len(raw_results)} crystals.")
+                # Sort from strongest to weakest in case of multiple crystals
+                await self._put_results(
+                    sorted(
+                        raw_results, key=lambda d: d[self.sort_key.value], reverse=True
+                    )
+                )
+
+            task = asyncio.create_task(_get_results())
+            try:
+                await asyncio.wait_for(asyncio.shield(task), self.timeout_s / 2)
+            except TimeoutError:
+                LOGGER.warning("Waited half of timeout for zocalo results - retrying")
+                await asyncio.wait_for(task, self.timeout_s / 2)
+
+        except TimeoutError:
+            LOGGER.warning("Timed out waiting for zocalo results!")
+            raise
+        finally:
+            self._kickoff_run = False
+
+    async def describe(self) -> dict[str, Descriptor]:
+        return OrderedDict(
+            [
+                (
+                    self._name + "-results",
+                    {
+                        "source": f"sim://{self._prefix}",
```

Comment:
Should: This is confusing as `_prefix` is never set so it gives all these with a source of `sim://`. Do we need to give something that matches a backend? If not I think a better source description would just be something like `zocalo_service`